include/tvm/target/target_id.h

```diff
@@ -0,0 +1,228 @@
+
+  /*! \brief Get the attribute map given the attribute name. */
+  template <typename ValueType>
+  static inline TargetIdAttrMap<ValueType> GetAttrMap(const String& attr_name);
+
+  TVM_DEFINE_NOTNULLABLE_OBJECT_REF_METHODS(TargetId, ObjectRef, TargetIdNode);
+
+ private:
+  static const AttrRegistryMapContainerMap<TargetId>& GetAttrMapContainer(const String& attr_name);
+  template <typename, typename>
+  friend class AttrRegistry;
+};
+
+/*!
+ * \brief Map<TargetId, ValueType> used to store meta-information about TargetId
+ * \tparam ValueType The type of the value stored in map
+ */
+template <typename ValueType>
+class TargetIdAttrMap : public AttrRegistryMap<TargetId, ValueType> {
+ public:
+  using TParent = AttrRegistryMap<TargetId, ValueType>;
+  using TParent::count;
+  using TParent::get;
+  using TParent::operator[];
+  explicit TargetIdAttrMap(const AttrRegistryMapContainerMap<TargetId>& map) : TParent(map) {}
+};
+
+/*!
+ * \brief Helper structure to register TargetId
+ * \sa TVM_REGISTER_TARGET_ID
+ */
+class TargetIdRegEntry {
+ public:
+  /*!
+   * \brief Register additional attributes to target_id.
+   * \param attr_name The name of the attribute.
+   * \param value The value to be set.
+   * \param plevel The priority level of this set,
+   *  an higher priority level attribute
+   *  will replace lower priority level attribute.
+   *  Must be bigger than 0.
+   *
+   *  Cannot set with same plevel twice in the code.
+   *
+   * \tparam ValueType The type of the value to be set.
+   */
+  template <typename ValueType>
+  inline TargetIdRegEntry& set_attr(const String& attr_name, const ValueType& value,
+                                    int plevel = 10);
+  /*!
+   * \brief Register a valid configuration option and its ValueType for validation
+   * \param key The configuration key
+   * \tparam ValueType The value type to be registered
+   */
+  template <typename ValueType>
+  inline TargetIdRegEntry& add_attr_option(const std::string& key);
+  /*! \brief Set name of the TargetId to be the same as registry if it is empty */
+  inline TargetIdRegEntry& set_name();
+  /*!
+   * \brief Register or get a new entry.
+   * \param name The name of the TargetId.
+   * \return the corresponding entry.
+   */
+  static TargetIdRegEntry& RegisterOrGet(const String& name);
+
+ private:
+  TargetId id_;
+  String name;
+
+  /*! \brief private constructor */
+  explicit TargetIdRegEntry(uint32_t reg_index) : id_(make_object<TargetIdNode>()) {
+    id_->index_ = reg_index;
+  }
+  /*!
+   * \brief update the attribute TargetIdAttrMap
+   * \param key The name of the attribute
+   * \param value The value to be set
+   * \param plevel The priority level
+   */
+  void UpdateAttr(const String& key, TVMRetValue value, int plevel);
+  template <typename, typename>
+  friend class AttrRegistry;
+};
+
+#ifndef TVM_STRINGIZE_DETAIL
+#define TVM_STRINGIZE_DETAIL(x) #x
```

Comment:
Do not need to duplicate the STRINGIZE macro, perhaps move everything to a common place

Reply:
I agree, but cannot find an existing file to put. What do you suggest?

Reply:
how about runtime/registry.h

Reply:
Sounds good, doing